torch/csrc/jit/passes/constant_propagation.cpp

```diff
@@ -63,28 +61,124 @@ void propagateNode(Node* n) {
   }
 }
 
+void lowerIf(Block *body, Node * n) {
+  auto graph = n->owningGraph();
+  WithInsertPoint insert_point_guard { n };
+
+  std::unordered_map<Value*, Value*> value_map;
+  auto get_value = [&](Value *v) {
+    auto it = value_map.find(v);
+    if (it != value_map.end())
+      return it->second;
+    return v;
+  };
+
+  for (Node *orig : body->nodes()) {
+    Node *clone = graph->insertNode(graph->createClone(orig, get_value));
+    for (size_t i = 0; i < orig->outputs().size(); ++i) {
+      value_map[orig->outputs()[i]] = clone->outputs()[i];
+    }
+  }
+  for (size_t i = 0; i < n->outputs().size(); ++i) {
+    n->outputs().at(i)->replaceAllUsesWith(get_value(body->outputs().at(i)));
+  }
+  // NB: destroy the node here, because it might contain side effects, like print
+  n->destroy();
+}
+
+bool isTrueConstant(Value *val) {
+  at::optional<bool> maybe_value = constant_as<bool>(val);
+  return maybe_value && *maybe_value;
+}
+
+void lowerIf(Node *n) {
+  if (isTrueConstant(n->input())) {
+    lowerIf(n->blocks()[0], n);
+  } else {
+    lowerIf(n->blocks()[1], n);
+  }
+}
+
+//returns true if the mutated variables are changed
+bool recomputeMutatedVariables(Node *n) {
+  JIT_ASSERTM(n->kind() == prim::If, "Only supported for If nodes");
+  std::unordered_set<Value*> mutated_variables;
+  for (Block * block : n->blocks()) {
+    for (Node *n : block->nodes()) {
+      for (size_t i = 0; i < n->outputs().size(); ++i) {
+        mutated_variables.insert(n->outputs()[i]);
+      }
+    }
+  }
+  auto true_block = n->blocks()[0];
+  auto false_block = n->blocks()[1];
+  auto initial_outputs = true_block->outputs().size();
+  for (size_t i = 0; i < true_block->outputs().size();) {
+    //neither block mutates output i
+    if (!mutated_variables.count(true_block->outputs()[i]) &&
```

Comment:
Isn't this whole function just equivalent to:
```
if(true_block->outputs()[i] == false_block->outputs()[i]) {
   // erase as an output
}
```

Reply:
yeaaaa good point lol